### PR TITLE
All websites dashboard - improve performance when thousands of websites

### LINKED ISSUE #6809: All websites dashboard - improve performance when thousands of websites
Currently on instance with really big number of websites (i.e. 20k+) Multisites dashboard tends to time out. It's caused by loading all sites into JS in one API call.
maybe there should be pagination implemented server-side so each API call only retrieves portion that would be displayed at one time to the user to speed up this view.

### PR DESCRIPTION
**Pull request by tsteur**
fixes #6809 
This improves the performance of the All Websites Dashboard. It includes improvements on several layers:
- Instead of fetching all sites into JS we do now only fetch the sites that we need to display. This means we will do a new request whenever someone sorts, searches, ...
- Before we always requested the `MultiSites.getAll` twice whenever the websites dashboard was loaded. One to get the current dashboard and one request to get the previous one (this was needed for a hover tooltip, the `past_data.total.nb_visits`). We do now only one request as the `MultiSites.getAll` already always requested the past data anyway. I do now expose this total visits value as well
- I made `MultiSites.getAll` on the server much faster. Eg before it took easily **> 300 seconds** to perform one `MultiSites.getAll` request when having 33k sites. This is now done in **less than 10 seconds**. If only a few sites are archived, and browser archiving is disabled, the performance **improved from > 60 seconds  to < 5 seconds**. Basically a problem was that we first created eg 100K DataTables, 1mio Rows, ... just to merge them afterwards together via `mergeChildren` to 1 DataTable and 33k Rows. This took a long time. Instead we want to directly build the merged dataTable. There are many changes in this pull request and each of them has actually a huge impact on MultiSites.getAll performance.

### CODE REVIEW

File: core/Archive/DataCollection.php [hunk 1]
```diff
@@ -303,9 +335,16 @@ private function createOrderedIndex($metadataNamesToIndexBy)
                 $indexKeyValues = array_keys($this->periods);
             }
 
-            foreach ($indexKeyValues as $key) {
-                $result[$key] = $this->createOrderedIndex($metadataNamesToIndexBy);
+            if (empty($metadataNamesToIndexBy)) {
+                foreach ($indexKeyValues as $key) {
+                    $result[$key] = array();
```
Comment (tsteur) [inline core/Archive/DataCollection.php]: NOTE to myself, use `array_fill_keys` or so

File: core/DataAccess/ArchiveSelector.php [hunk 2]
```diff
@@ -153,9 +153,13 @@ public static function getArchiveIds($siteIds, $periods, $segment, $plugins)
...
         }
 
+        foreach ($siteIds as $index => $siteId) {
+            $siteIds[$index] = (int) $siteId;
+        }
+
         $getArchiveIdsSql = "SELECT idsite, name, date1, date2, MAX(idarchive) as idarchive
                                FROM %s
-                              WHERE idsite IN (" . Common::getSqlStringFieldsArray($siteIds) . ")
+                              WHERE idsite IN (" . implode(',', $siteIds) . ")
```
Comment (tsteur) [inline core/DataAccess/ArchiveSelector.php]: This improved one query from 3.5 seconds to < 100ms

File: core/Site.php [hunk 3]
```diff
@@ -99,24 +123,6 @@ protected static function setSite($idSite, $infoSite)
...
-         * **Example**
-         *
-         *     Piwik::addAction('Site.setSite', function ($idSite, &$info) {
-         *         $info['name'] .= " (original)";
-         *     });
-         *
-         * @param int $idSite The ID of the website entity that will be modified.
-         * @param array $infoSite The website entity. [Learn more.](/guides/persistence-and-the-mysql-backend#websites-aka-sites)
-         */
-        Piwik::postEvent('Site.setSite', array($idSite, &$infoSite));
```
Comment (tsteur) [inline core/Site.php]: Here we will break API, it was only there for `MetaSites` plugin but in theory other plugins could have used it as well. We spend I think > 7 seconds just here in triggering this event. We decided to deprecate it. I will need to adjust MetaSites plugin, already started doing it.

File: plugins/MultiSites/API.php [hunk 4]
```diff
@@ -221,6 +225,10 @@ private function buildDataTable($sitesToProblablyAdd, $period, $date, $segment,
         $apiECommerceMetrics = array();
         $apiMetrics = API::getApiMetrics($enhanced);
         foreach ($apiMetrics as $metricName => $metricSettings) {
+            if (!empty($showColumns) && !in_array($metricName, $showColumns)) {
```
Comment (tsteur) [inline plugins/MultiSites/API.php]: Do not load metrics that we do not need...

File: plugins/MultiSites/Columns/Metrics/EcommerceOnlyEvolutionMetric.php [hunk 5]
```diff
@@ -36,13 +36,13 @@ public function compute(Row $row)
 
         // if the site this is for doesn't support ecommerce & this is for the revenue_evolution column,
         // we don't add the new column
-        if (($currentValue === false
-            || !$this->isRevenueEvolution)
-            && !Site::isEcommerceEnabledFor($row->getColumn('label'))
-        ) {
-            $row->deleteColumn($columnName);
+        if ($currentValue === false || !$this->isRevenueEvolution) {
+            $idSite = $row->getMetadata('idsite');
```
Comment (tsteur) [inline plugins/MultiSites/Columns/Metrics/EcommerceOnlyEvolutionMetric.php]: Label might be already replaced with the Site name. We need to use the idsite metadata instead which always works

File: plugins/MultiSites/Controller.php [hunk 6]
```diff
@@ -40,6 +45,34 @@ public function standalone()
         return $this->getSitesInfo($isWidgetized = true);
     }
 
+    public function getAllWithGroups()
+    {
+        Piwik::checkUserHasSomeViewAccess();
```
Comment (tsteur) [inline plugins/MultiSites/Controller.php]: All this is in a controller and not API. It is not an API method as it only does some formatting etc for our dashboard widget.

File: core/Archive/DataTableFactory.php [hunk 7]
```diff
@@ -157,21 +161,65 @@ public function useSubtable($idSubtable)
      */
     public function make($index, $resultIndices)
     {
+        $keyMetadata = $this->getDefaultMetadata();
```
Comment (tsteur) [inline core/Archive/DataTableFactory.php]: A lot of time was spent in setting and transforming metadata. This is much faster now. Eg before we did set a table metadata like `array('period' => '2012-12...', 'site' => 5)` and they were afterwards "transformed" to `array('period' => Period, 'site' => new Site())`.

File: core/DataTable.php [hunk 8]
```diff
@@ -1460,12 +1459,13 @@ public function addRowsFromSimpleArray($array)
                 // it cannot be lost during the conversion. Because we are not able to handle properly
                 // this key, we throw an explicit exception.
                 if (is_string($key)) {
-                    throw $e;
+                    // we define an exception we may throw if at one point we notice that we cannot handle the data structure
+                    throw new Exception(sprintf($exceptionText, var_export($array, true)));
```
Comment (tsteur) [inline core/DataTable.php]: Only call `var_export` when needed... we called it > 100k times although it was not needed

File: plugins/MultiSites/API.php [hunk 9]
```diff
@@ -229,25 +237,11 @@ private function buildDataTable($sitesToProblablyAdd, $period, $date, $segment,
             }
         }
 
-        // get the data
-        // $dataTable instanceOf Set
-        $dataTable = $archive->getDataTableFromNumeric($fieldsToGet);
```
Comment (tsteur) [inline plugins/MultiSites/API.php]: Before we built a huge datable (map) and then merged them afterwards see `mergeChildren`, this was very expensive in terms of time.

File: core/Archive/DataTableFactory.php [hunk 10]
```diff
@@ -267,23 +312,26 @@ private function makeIndexedByRecordNameDataTable($blobRow)
...
-            $keyMetadata[$resultIndex] = $label;
-
-            if (empty($resultIndices)) {
-                $newTable = $this->createDataTable($value, $keyMetadata);
-            } else {
+            if ($resultIndex === DataTableFactory::TABLE_METADATA_SITE_INDEX) {
+                $keyMetadata[$resultIndex] = new Site($label);
+            } elseif ($resultIndex === DataTableFactory::TABLE_METADATA_PERIOD_INDEX) {
+                $keyMetadata[$resultIndex] = $this->periods[$label];
+            }
```
Comment (diosmosis) [inline core/Archive/DataTableFactory.php]: Small code clarity item: I think this might be clearer in a private method, ie, `$keyMetadata[$resultIndex] = $this->createTableIndexMetadata($resultIndex, $label);`

File: core/Archive/DataTableFactory.php [hunk 11]
```diff
@@ -157,21 +161,62 @@ public function useSubtable($idSubtable)
...
+     * eg `MultiSites.getAll`. Was done as part of https://github.com/piwik/piwik/issues/6809
+     *
+     * @param array $index @see DataCollection
+     * @param array $resultIndices an array mapping metadata names with pretty metadata labels.
+     *
+     * @return DataTable|DataTable\Map
+     * @throws \Exception
+     */
+    public function makeMerged($index, $resultIndices)
+    {
```
Comment (diosmosis) [inline core/Archive/DataTableFactory.php]: I think instead of makeMerged + makeMergedTableWithPeriodAndSiteIndex/makeMergedWithSiteIndex, this method could be made generic w/ a `makeMerged($index, $resultIndices, $mergeIndices)`. I haven't coded an implementation so I don't know if there are any hidden problems, but I think it work by:

1. creating a nested DataTable\Map using values in $resultIndices (unless it's empty)
2. then creating individual tables & setting columns using the keys in $mergeIndices

Generic code means cleaner, reasoned code, and thus less technical debt. Hopefully, it's possible w/o optimization trade-offs.

File: core/DataAccess/ArchiveSelector.php [hunk 12]
```diff
@@ -172,7 +176,7 @@ public static function getArchiveIds($siteIds, $periods, $segment, $plugins)
         foreach ($monthToPeriods as $table => $periods) {
             $firstPeriod = reset($periods);
 
-            $bind = array_values($siteIds);
+            $bind = array();
```
Comment (diosmosis) [inline core/DataAccess/ArchiveSelector.php]: Is this variable still necessary?

File: core/DataTable.php (hunk 8 above)
Comment (diosmosis) [inline core/DataTable.php]: I think this (and the throw statement below) can be moved to a separate method.

File: core/DataTable/Filter/ColumnCallbackAddMetadata.php [hunk 13]
```diff
@@ -45,10 +45,6 @@ public function __construct($table, $columnsToRead, $metadataToAdd, $functionToA
     {
         parent::__construct($table);
 
-        if (!is_array($columnsToRead)) {
-            $columnsToRead = array($columnsToRead);
-        }
-
         $this->columnsToRead      = $columnsToRead;
```
Comment (diosmosis) [inline core/DataTable/Filter/ColumnCallbackAddMetadata.php]: Could you tell me how much of an improvement handling the single columnsToRead case has? Also, does this optimize something in MultiSites.getAll?

File: lang/en.json [hunk 14]
```diff
@@ -31,7 +31,7 @@
         "ChooseWebsite": "Choose website",
         "ClickHere": "Click here for more information.",
         "ClickToChangePeriod": "Click again to change the period.",
-        "Close": "Close",
+        "ClickToSearch": "Click to search",
```
Comment (diosmosis) [inline lang/en.json]: Is this change intentional?

File: plugins/MultiSites/Dashboard.php [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+use Piwik\DataTable\Row\DataTableSummaryRow;
+use Piwik\Plugins\API\ProcessedReport;
+use Piwik\Site;
+use Piwik\View;
+
+/**
+ * Fetches and formats the response of `MultiSites.getAll` in a way that it can be used by the All Websites AngularJS
+ * widget. Eg sites are moved into groups if one is assigned, stats are calculated for groups, etc.
+ */
+class Dashboard
```
Comment (diosmosis) [inline plugins/MultiSites/Dashboard.php]: This is the class that's responsible for applying the limit/offset/sort to the MultiSites.getAll result, right?

File: plugins/MultiSites/angularjs/dashboard/dashboard-model.service.js [hunk 16]
```diff
@@ -46,153 +31,52 @@
...
-
-            for (var index = 0; index < model.sites.length; index++) {
-                var site = model.sites[index];
-                if (site && site.isGroup) {
-                    numSites += site.sites.length;
-                }
-            }
-
-            return numSites;
+            return model.numberOfSites; // todo
```
Comment (diosmosis) [inline plugins/MultiSites/angularjs/dashboard/dashboard-model.service.js]: Does this todo need to be taken care of?

File: tests/UI/specs/MultiSites_spec.js [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+
+describe("MultiSitesTest", function () {
+    this.timeout(0);
+
+    var generalParams = 'idSite=1&period=year&date=2012-08-09';
+    var selector = '.pageWrap,.expandDataTableFooterDrawer';
+
+    beforeEach(function () {
+        delete testEnvironment.configOverride;
+        testEnvironment.testUseRegularAuth = 0;
```
Comment (diosmosis) [inline tests/UI/specs/MultiSites_spec.js]: I don't think these two settings should be necessary.

File: core/DataAccess/ArchiveSelector.php (hunk 12 above)
Comment (tsteur) [inline core/DataAccess/ArchiveSelector.php]: Yes

File: core/DataTable/Filter/ColumnCallbackAddMetadata.php (hunk 13 above)
Comment (tsteur) [inline core/DataTable/Filter/ColumnCallbackAddMetadata.php]: I don't remember the exact time but it was certainly an improvement worth it otherwise I wouldn't have done it. It is actually no longer used after I replaced it afterwards with `MetadataCallbackAddMetadata`. I reckon we can still leave it there for improved performance in other use cases.

File: lang/en.json (hunk 14 above)
Comment (tsteur) [inline lang/en.json]: Nope, reckon it got lost while rebasing or so.

File: plugins/MultiSites/Dashboard.php (hunk 15 above)
Comment (tsteur) [inline plugins/MultiSites/Dashboard.php]: It's responsible to prepare / fetch / format data for the all websites dashboard. Therefore it does limit/offset/sort yes

File: plugins/MultiSites/angularjs/dashboard/dashboard-model.service.js (hunk 16 above)
Comment (tsteur) [inline plugins/MultiSites/angularjs/dashboard/dashboard-model.service.js]: Nope, will remove it

File: tests/UI/specs/MultiSites_spec.js (hunk 17 above)
Comment (tsteur) [inline tests/UI/specs/MultiSites_spec.js]: Can remove it and see if it still works

File: core/Archive/DataTableFactory.php (hunk 11 above)
Comment (tsteur) [inline core/Archive/DataTableFactory.php]: I won't make it generic when it's currently only used in one use case and maybe even removed in a few weeks or months again (when we archive `MultiSites.getAll` as a blob for even better performance). We can certainly refactor it once we need to pass `$mergeIndices` or something else. Generic code when not needed isn't always good and often results in over-engineered code etc and things we do not need. I don't think it would get much easier to read here rather more abstract and even harder to understand. I would prefer, when this method needs to be used in another case, to let it evolve and get it more generic in a way it is actually needed.

File: core/DataTable/Filter/ColumnCallbackAddMetadata.php (hunk 13 above)
Comment (tsteur) [inline core/DataTable/Filter/ColumnCallbackAddMetadata.php]: Found a profile from a run see screenshot. The interesting part is the exclusive metric for that function which took 2.6seconds although only called 3 times. This is a long time and got done quite a bit by this tweak but don't remember how much. Ignore the 1.3% wall time. This was before optimizing anything. Once I tweaked other parts this become like 10% or 20% wall time

columnaddcallback

File: core/Archive/DataTableFactory.php (hunk 10 above)
Comment (tsteur) [inline core/Archive/DataTableFactory.php]: done

File: core/Archive/DataTableFactory.php (hunk 11 above)
Comment (diosmosis) [inline core/Archive/DataTableFactory.php]: If you don't want to make this addition to core generic (and thus not tied to a specific use case in a single plugin), can you do one of the following things:

1) _(quickest)_ Mark Archive::getDataTableNumericMerged or whatever it is called as `@internal` and create an issue about removing it and archiving MultiSites.getAll. Please leave a comment that says the Archive & DataTableFactory methods should be dealt w/ some other way if this new ticket is considered invalid and closed w/o being implemented.
2) Move the logic to the MultiSites plugin. If another plugin needs something like it, then it can be moved to core.

File: core/DataTable/Filter/ColumnCallbackAddMetadata.php (hunk 13 above)
Comment (diosmosis) [inline core/DataTable/Filter/ColumnCallbackAddMetadata.php]: I don't like the idea of explicitly including redundant code in core. If this is not important to the optimizations, I would suggest moving it to a new issue where someone can hopefully find a reason for the speed increase & find a clean fix. I'll leave it to @mattab to decide whether it's worth it to include this debt now.

File: plugins/MultiSites/Dashboard.php (hunk 15 above)
Comment (diosmosis) [inline plugins/MultiSites/Dashboard.php]: Can you add a note to the class docs? I found it hard to determine what it did differently than simply querying MultiSites.getAll.

Comment (tsteur) [inline plugins/MultiSites/Dashboard.php]: There is already one or not? What is missing?

File: core/Archive/DataTableFactory.php (hunk 11 above)
Comment (tsteur) [inline core/Archive/DataTableFactory.php]: I marked it as internal but I won't create an issue for archiving `MultiSites.getAll`. If MultiSites needs to perform better an issue will be created naturally. Archiving `MultiSites.getAll` will be most likely the solution but that's not necessarily the case.

There is simply no reason to make it more generic currently. This will come as well naturally by changed requirements (we don't know those requirements yet). So there is no reason to deal with those methods some other way until needed somewhere else or whatever. It's just important to refactor and change it when needed, or to remove it again if no longer needed.

File: core/DataTable/Filter/ColumnCallbackAddMetadata.php (hunk 13 above)
Comment (tsteur) [inline core/DataTable/Filter/ColumnCallbackAddMetadata.php]: I can remove it if you want, I don't care about it. It's only a bit of duplicated code and right next to each other not across files or functions or whatever. In general duplicated code is not good, yes, that's why I added that comment initially but sometimes it's ok. Anyway, I will just remove it

**Comment by mattab** [comment]
Needs a rebase & then +1 to merge it before the beta/RC

File: plugins/MultiSites/Dashboard.php (hunk 15 above)
Comment (diosmosis) [inline plugins/MultiSites/Dashboard.php]: The class doc says nothing about sorting/offset/limit (or in other words supporting pagination), yet that appears to be the main reason to use Dashboard.php and not MultiSites.getAll directly. It's important in quickly understanding what the class is for.

Comment (tsteur) [inline plugins/MultiSites/Dashboard.php]: Sorting, offset, limit is not a main reason to use Dashboard.php it only does it along other things. For those things I could use MultiSites.getAll directly. What it is important for is to move sites into groups, to calculate stats for groups etc. This is described in the doc.
